ambry-router/src/main/java/com.github.ambry.router/AdaptiveOperationTracker.java

```diff
@@ -123,14 +124,14 @@ public boolean hasNext() {
     @Override
     public void remove() {
       replicaIterator.remove();
-      if (inflightCount >= parallelism) {
+      if (inflightCount >= parallelism && unexpiredRequestSendTimes.size() > 0) {
         // we are here because oldest request is past due
-        Map.Entry<ReplicaId, Long> oldestEntry = unexpiredRequestSendTimes.entrySet().iterator().next();
-        expiredRequestSendTimes.put(oldestEntry.getKey(), oldestEntry.getValue());
+        Map.Entry<ReplicaId, Pair<Boolean, Long>> oldestEntry = unexpiredRequestSendTimes.entrySet().iterator().next();
+        expiredRequestSendTimes.put(oldestEntry.getKey(), oldestEntry.getValue().getSecond());
         unexpiredRequestSendTimes.remove(oldestEntry.getKey());
         pastDueCounter.inc();
       }
-      unexpiredRequestSendTimes.put(lastReturned, time.milliseconds());
+      unexpiredRequestSendTimes.put(lastReturned, new Pair<>(false, time.milliseconds()));
```

Comment:
Is it possible and if so, does it make more sense to do this (and incrementing `inFlightCount`) in `next()` and not in `remove()`. Seems like that is the right place to do it and you would not need to keep `lastReturned`.